### ISSUE TITLE
Updates can't actually be queued and postponed indefinitely

### ISSUE THREAD

**Issue by KtorZ** [input-output-hk/cardano-wallet-legacy#151]
| Release | Operating System       | Cause |
| ---     | ---                    | ---   |
| next    | Windows & OSX & Linux) | Code |

# Context
 
 

Here below is describe how updates are currently handled in `cardano-sl` (in
cunjunction with `cardano-wallet` still used as a middle-man for communicating
between `Daedalus` and its underlying node).

At epoch boundaries, node may extract and process confirmed update proposals
received from the network. A worker running asynchronously regularly checks for 
updates and eventually triggers the `downloadUpdate` function below:

**Pos.Network.Update.Download**  
```hs
downloadUpdate :: forall ctx m . UpdateMode ctx m => ConfirmedProposalState -> m ()
```

In its normal mode, this function does two major things (beside a bunch of checks 
on the update's version and target system):

- It downloads the corresponding update and write it to disk at a given location.
  (the location depends on a CLI option given at runtime). 

- It put the corresponding `ConfirmedProposalState` in an `MVar` shared with the 
  wallet backend. 

**Pos.DB.Update.Context**  
```hs
data UpdateContext = UpdateContext
    {
    -- | A semaphore which is unlocked when update data is downloaded and
    -- ready to apply.
      ucDownloadedUpdate :: !(MVar ConfirmedProposalState)

    -- | A lock which allows only one thread to download an update.
    , ucDownloadLock     :: !(MVar ())

    -- | In-memory state of update-system-as-block-component.
    , ucMemState         :: !MemVar
    }
```

The wallet backend is the only listener on this `ucDownloadedUpdate` `MVar` which it polls
regularly checking for available updates. Pretty much as soon as the `MVar` is filled, it
is taken by the wallet which stores the corresponding update in a queue persisted in acid-state.

From there, the frontend which is regularly polling the wallet backend about the next available 
update get informed that a new update is available in the queue. The user is prompted
with a choice of either:

- Applying the update by shutting down the node immediately with an exit code 20 (which tells the launcher to consequently restart the node) 
- Postponing the shut down, which will apply the update only when the node restarts

Regardless of the choice, the update is popped out of the wallet's queue will be applied when 
the node restarts. 

> NOTE: 
> For 'core' nodes with no wallet running, the behavior is slightly different. An extra worker
> is started which role is to restart the node immediately when an update is available.
> 
> **Pos.Worker.Update**  
> ```hs
> updateTriggerWorker :: UpdateMode ctx m => Diffusion m -> m ()
> updateTriggerWorker = \_ -> do
>     logInfo "Update trigger worker is locked"
>     void $ takeMVar . ucDownloadedUpdate =<< view (lensOf @UpdateContext)
>     triggerShutdown
> ```

# Steps to Reproduce

If we wait long enough without restarting, the node may not be able to download the next update at all.

When a new update arrives, the node starts by checking if the `ucDownloadedUpdate` `MVar` is empty, 
and if so, assume the new update can be downloaded. 

**Pos.Network.Update.Download**
```hs
downloadUpdate :: forall ctx m . UpdateMode ctx m => ConfirmedProposalState -> m ()
downloadUpdate cps = do
    [...]
        tryReadMVar downloadedUpdateMVar >>= \case
            Nothing -> do
                updHash <- getUpdateHash cps
                installed <- isUpdateInstalled updHash
                if installed
                    then onAlreadyInstalled
                    else downloadUpdateDo updHash cps
```

Yet, in `downloadUpdateDo` below, the code throws when there's already an existing 
update downloaded (which remains unless applied). As a result, the update 

**Pos.Network.Update.Download**  
```hs
downloadUpdateDo :: UpdateMode ctx m => Hash Raw -> ConfirmedProposalState -> m ()
downloadUpdateDo updHash cps@ConfirmedProposalState {..} = do
    [...]
    res <- handleAny handleErr $ runExceptT $ do
        [...]
        updPath <- views (lensOf @UpdateParams) upUpdatePath
        whenM (liftIO $ doesFileExist updPath) $
            throwError "There's unapplied update already downloaded"
    [...]
  where
    handleErr e =
        Left (pretty e) <$ reportOrLogW "Update downloading failed: " e
```

From here, there's no way to recover the missed update and a node has to either
rollback manually some blocks, or, to wait for the next update hoping it'll have
restarted in the meantime.

Note that, if we were to leave the `ucDownloadedUpdate` `MVar` full by not taking its content, 
the node would also ignore the update:

**Pos.Network.Update.Download**  
```hs
downloadUpdate :: forall ctx m . UpdateMode ctx m => ConfirmedProposalState -> m ()
downloadUpdate cps = do
    [...]
        tryReadMVar downloadedUpdateMVar >>= \case
            [...]
            Just existingCPS -> onAlreadyDownloaded existingCPS
  where
    [...]
    onAlreadyInstalled =
        logInfo $ sformat ("We won't download an update for proposal "
            %build%", because it's already installed") proposalToDL
```

So. There's definitely something wrong in the way we currently manage updates through the wallet code.
If we still want to support this postponing feature, we may either force a restart after some time, or
review how updates are handled all-in-all.

---

# Resolution Plan

Discussing with frontend folks, we agreed on the following:

- /api/v1/postpone-update is removed.
- /api/v1/apply-update is renamed to /api/v1/restart-node and now only restarts the node (so, shutdown with exit code 20 that gets picked up by the launcher). We also preserve the faulty behavior injection.
- /api/v1/next-update will now return 200 with update information when there's an available update and until it gets applied (whereas you could "dismiss" this currently by calling postpone-update). It will still return a 404 when there's no update.
So basically, `next-update` becomes pretty much just a getter on the underlying ucDownloadedUpdate MVar.

This change will happen on the new Node API, and the wallet will act as a reverse proxy for this behavior.

# PR

| Number   | Base      |
| ---      | ---       |
| #?       | `develop` |

# QA

**Comment by parsonsmatt** [comment]
I believe we can close this issue now, as the node and wallet have udpated behavior.

**Comment by piotr-iohk** [comment]
Hey @parsonsmatt, can you provide info in the QA section? In particular, in the _Resolution Plan_ I see there was a plan to make changes to API endpoints:
`/api/v1/postpone-update`, `/api/v1/apply-update`, `/api/v1/next-update` (btw. shouldn't it be `/api/internal/...` instead of `/api/v1/...`?)
If that's the case then there's an opportunity for adding some integration tests for newly added/modified endpoints? :)

If the original _Resolution Plan_ changed can you please provide such info in the ticket?
For the time being I will put the item back into _In progress_ due to the lack of this information.
CC: @KtorZ.

**Comment by parsonsmatt** [comment]
Hmm -- an integration test seems like it would be difficult to orchestrate,  because it requires that the underlying node processes see an update and respond appropriately.

As far as the *wallet* is concerned, it is delegating to the Node API. The code on `develop` is this:

```haskell
nextUpdate :: NodeHttpClient -> Handler (APIResponse WalletSoftwareVersion)
nextUpdate nc = do
    emUpd <- liftIO . runExceptT $ NodeClient.getNextUpdate nc
    case emUpd of
        Left err  ->
            handleNodeError err
        Right (NodeClient.V1 upd) ->
            single <$> pure (WalletSoftwareVersion upd)

applyUpdate :: NodeHttpClient -> Handler NoContent
applyUpdate nc = do
    enc <- liftIO . runExceptT $ NodeClient.restartNode nc
    case enc of
        Left err ->
            handleNodeError err
        Right () ->
            pure NoContent

-- | This endpoint has been made into a no-op.
postponeUpdate :: Handler NoContent
postponeUpdate = pure NoContent
```

So `postponeUpdate` is not removed, but it is turned into a no-op. The other two update endpoints delegate to the Node, and if the node behavior is correct, then the wallet behavior is also correct.

The code for the web/API section of the updates on the node API is here:

```haskell
--------------------------------------------------------------------------------
-- Updates
--------------------------------------------------------------------------------

-- | Handler
restartNode :: ShutdownContext -> Handler NoContent
restartNode shutdownCtx = liftIO $ do
    doFail <- testLogFInject (_shdnFInjects shutdownCtx) FInjApplyUpdateNoExit
    unless doFail (runReaderT triggerShutdown shutdownCtx)
    pure NoContent

-- | This endpoint does a 404 unless there is an update available. If an update
-- is available, it returns the 'SoftwareVersion' for that update.
getNextUpdate :: UpdateContext -> Handler (APIResponse (V1 SoftwareVersion))
getNextUpdate uc = do
    mproposalState <- tryReadMVar (ucDownloadedUpdate uc)
    single <$> case mproposalState of
        Just proposalState ->
            pure (V1 (upSoftwareVersion (cpsUpdateProposal proposalState)))
        Nothing ->
            throwError err404
```

So we could test the `Handler` for `getNextUpdate` by providing a fake `UpdateContext`, calling `getNextUpdate` in test code, and verify that it returns `Right softwareVersion` when the `MVar` is filled and `Left err404` when the `MVar` is empty. This doesn't strike me as a high value test, but it's also not hard to write.

`restartNode` can be tested whether calling that function writes to the `TVar`. But there is 0 business logic in that handler, so that also doesn't strike me as a high value test.

**Comment by piotr-iohk** [comment]
Hey @parsonsmatt. Thanks for the answer!
As for the tests for `restartNode` and `getNextUpdate` - I'm for implementing those tests. My opinion is actually they would be good tests to have. I think it's better to have even simple tests than no tests at all ;) At least it will add some additional code coverage, even if it is not excercising any significant business scenario...

As for the `nextUpdate` and `applyUpdate` - got it (that it is not easy to implement meaningful business scenarios here). In such a case maybe the test could be just a simple request to both endpoints?
e.g. `nextUpdate` should return error 404 if there is no update on the node (and that is something)
`applyUpdate` is doing `NodeClient.restartNode` so no update proposal is needed (I'm guessing:))

BTW. since `applyUpdate` is doing `NodeClient.restartNode` shouldn't it be renamed to `restartNode` for better consistency? I'm referring also to resolution plan... 
> /api/v1/apply-update is renamed to /api/v1/restart-node

Additionally https://input-output-hk.github.io/cardano-wallet/#tag/Internal does not seem to be updated? Still there are references to `postpone-update` and `apply-update`

**Comment by KtorZ** [comment]
I agree with @parsonsmatt on the "not so high value" in testing that `getNextUpdate` returns something when the MVar is full and 404 otherwise while providing a mock context to do so. Being able to submit update in our integration test and make sure they're correctly reported would be the way to go yet, this is much more complicated to orchestrate and is pretty much out-of-scope for this ticket (but perhaps, worth a ticket of its own). 

About the restartNode, this _could_ be fairly easy to test and is worth testing; especially because we already had issues with that in past releases. This could even fit in our testing scenario. We make a restart request, we then poll the node until it's back up, controlling that it was down for a while. The node should restart and reconnect after a few seconds. 

Ideally, we would like to control the exit code and make sure it's 20, but that's a bit harder to instrument with our scenarios (though we have access to the async handle, so I guess there are ways to retrieve the exit code of a particular process? I actually wonder if, restarting a the node wouldn't restart the entire integration tests. We aren't starting new processes here, but rather forking the main thread. Therefore, is we exit the current process, all threads will die I believe? In the latter case, that could become quite difficult to instrument without broader changes in the integration cluster (which will be required anyway once we integrate with the rust nodes)).  

My suggestion for now would be to draft some manual instructions on how to test that (i.e. how to have a fake update in the system, and what calls to make to the API to check that it appears and get applied correctly). 

> BTW. since applyUpdate is doing NodeClient.restartNode shouldn't it be renamed to restartNode for better consistency? I'm referring also to resolution plan... 

Agreed and that was part of the initial plan. Since these are internal endpoints, that's all fine. I've already discussed the matter with Daedalus' folks and that's all good for them as well.

**Comment by piotr-iohk** [comment]
OK. To sum up then:

 - manual procedure for _proper_ test for `nextUpdate` to be done as part of this ticket (here? -> https://github.com/input-output-hk/cardano-wallet/tree/develop/test/manual)
 - create separate ticket for implementing integration test for `nextUpdate` (based on the manual procedure)

 - `restartNode` - ~make attempt to~ implement such integration test under this ticket (~if it turns out to be more problematic and will require more changes to the cluster, then perhaps create separate ticket?~)

 - update API doc (remove `/api/internal/postpone-update` if that's the case, say `/api/internal/apply-update` is only restarting the node - if that's the case)

@parsonsmatt , @KtorZ - does the above sound reasonable to you? :)

---
btw. 
>Since these are internal endpoints, that's all fine. I've already discussed the matter with Daedalus' folks and that's all good for them as well.

OK, thanks! 
(wouldn't be asking if this was in the ticket ;) )
Still, it is a bit counter-intuitive to me (and perhaps anyone who's going to read API doc)...

**Comment by KtorZ** [comment]
@piotr-iohk Yes, although for the third point, I'd go right away for the ticket option. This is going to be non-trivial to instrument in automated tests.

**Comment by KtorZ** [comment]
>  Still, it is a bit counter-intuitive to me (and perhaps anyone who's going to read API doc)...

We have several options here: only show the internal endpoints documentation when ran under special flags (like a "dev-mode") for Daedalus to run. Some API endpoints used to be only accessible via `--debug` but this wasn't ideal because `--debug` also deactivate TLS :/ ... 
And, we do need the endpoints in production mode, we may however remove them from the documentation. 

Or, at the very least, add a proper description for the internal group explaining they're only intended to be used by power users like Daedalus. I thought it was the case already.

(edit: ticket to be completed: https://github.com/input-output-hk/cardano-wallet/issues/228)

**Comment by piotr-iohk** [comment]
>Or, at the very least, add a proper description for the internal group explaining they're only intended to be used by power users like Daedalus. I thought it was the case already.

After thinking about it I come to conclusion I'd be for adding proper description for the internal group (explaining that these are only to be used by Daedalus etc.) rather than hiding. I guess that's more in the spirit of _open source_ :) (that's not a strong opinion though, so if you guys think otherwise, I'm ok:) )

In any case, putting the ticket back to _in progress_ for the remaining items -> https://github.com/input-output-hk/cardano-wallet/issues/151#issuecomment-454721959

**Comment by piotr-iohk** [comment]
@KtorZ do we want to delegate the remaining actions to separate ticket then, or...?
>- manual procedure for proper test for nextUpdate to be done as part of this ticket (here? -> >https://github.com/input-output-hk/cardano-wallet/tree/develop/test/manual)
>
>- create separate ticket for implementing integration test for nextUpdate (based on the manual >procedure)
>
>- restartNode - make attempt to implement such integration test under this ticket (if it turns out >to be more problematic and will require more changes to the cluster, then perhaps create separate >ticket?)
>
>- update API doc (remove /api/internal/postpone-update if that's the case, say /api/internal/apply->update is only restarting the node - if that's the case)

**Comment by KtorZ** [comment]
I'd say: yes. Especially because the last point will cause a breaking change for Daedalus and that's not something we want to do at the edge of a release. So, we'll do it for the next one.

**Comment by piotr-iohk** [comment]
#296
